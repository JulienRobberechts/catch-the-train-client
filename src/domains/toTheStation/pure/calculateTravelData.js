import moment from "moment";
import getDelayStatus from "./getDelayStatus";

export default function calculateTravelData({
  nowTime,
  targetTime,
  travelDuration,
  waitingDuration,
  onTimeMarginDelaySeconds
}) {
  const targetDuration = moment.duration(targetTime.diff(nowTime));

  // positive will be early and negative late
  const delayDuration = targetDuration.clone();
  delayDuration.subtract(travelDuration);
  delayDuration.subtract(waitingDuration);

  const delayStatus = getDelayStatus(delayDuration, onTimeMarginDelaySeconds);

  return {
    targetDuration,
    delayDuration,
    delayStatus
  };
}
