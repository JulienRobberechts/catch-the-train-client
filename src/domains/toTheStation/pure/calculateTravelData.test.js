import moment from "moment";
import calculateTravelData from "./calculateTravelData";
import each from "jest-each";

describe("train helpers", () => {
  describe("calculateTravelData", () => {
    test("should identify an late delay", () => {
      const nowTime = moment.parseZone("2020-03-10T09:00:00+01:00");
      const departure = { departureTime: "2020-03-10T09:11:00+01:00" };
      const travelDurationSeconds = 8 * 60;
      const waitingDelaySeconds = 120;
      const onTimeMarginDelaySeconds = 30;

      const result = calculateTravelData({
        nowTime,
        departure,
        travelDurationSeconds,
        waitingDelaySeconds,
        onTimeMarginDelaySeconds
      });

      expect(result).toEqual(expect.anything());
      expect(result).toMatchSnapshot();

      expect(result.delayDuration.asSeconds()).toEqual(60);
    });
  });
});
