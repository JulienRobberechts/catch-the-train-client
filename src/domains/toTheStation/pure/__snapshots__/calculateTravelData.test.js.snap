// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`train helpers calculateTravelData should identify an late delay 1`] = `
Object {
  "delayDuration": "PT1M",
  "delayStatus": "early",
  "nowTime": "2020-03-10T08:00:00.000Z",
  "route": undefined,
  "targetDuration": "PT11M",
  "targetTime": "2020-03-10T08:11:00.000Z",
  "travelDuration": "PT8M",
  "waitingDuration": "PT2M",
}
`;
